Index: zfs-2.1.2/lib/libefi/rdwr_efi.c
===================================================================
--- zfs-2.1.2.orig/lib/libefi/rdwr_efi.c
+++ zfs-2.1.2/lib/libefi/rdwr_efi.c
@@ -157,7 +157,7 @@ struct dk_map2  default_vtoc_map[NDKMAP]
 
 #if defined(i386) || defined(__amd64) || defined(__arm) || \
     defined(__powerpc) || defined(__sparc) || defined(__s390__) || \
-    defined(__mips__) || defined(__rv64g__)
+    defined(__mips__) || defined(__rv64g__) || defined(__loongarch64)
 	{	V_BOOT,		V_UNMNT	},		/* i - 8 */
 	{	V_ALTSCTR,	0	},		/* j - 9 */
 
Index: zfs-2.1.2/lib/libspl/include/sys/isa_defs.h
===================================================================
--- zfs-2.1.2.orig/lib/libspl/include/sys/isa_defs.h
+++ zfs-2.1.2/lib/libspl/include/sys/isa_defs.h
@@ -233,6 +233,17 @@ extern "C" {
 
 #define	_SUNOS_VTOC_16
 
+/* loongarch specific defines */
+#elif defined(__loongarch64)
+
+#if !defined(_LP64)
+#define	_LP64
+#endif
+
+#define	_ZFS_LITTLE_ENDIAN
+
+#define	_SUNOS_VTOC_16
+
 #else
 /*
  * Currently supported:
